### PR TITLE
Add skip-get-modules flag to support isolated envs

### PR DESCRIPTION
**Pull request by foadnh**
**Description:** Add `skip-get-modules` flag to the builder.

Isolated build environments like [Bazel](https://bazel.build/) or some CIs might not have access to the public internet.
These environments have different ways to inject dependencies, and there is no need to get them directly from the internet during the builder execution time.
~~Also, if we skip compilation (an existing feature) and downloading modules (this feature), there is no need for Go binary anymore.~~ Move to a separate PR.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,9 @@
   - `processorhelper.New[Traces|Metrics|Logs]ProcessorWithCreateSettings`
   - `component.NewExtensionFactoryWithStabilityLevel`
 
+### 💡 Enhancements 💡
+
+- Add `skip-get-modules` builder flag to support isolated environment executions (#6009)
```
Comment (dmathieu) [inline CHANGELOG.md:14]: I think you need an extra new line after your new entry and before the previous release version.

File: cmd/builder/internal/builder/config.go [hunk 2]
```diff
@@ -89,15 +90,17 @@ func NewDefaultConfig() Config {
 // Validate checks whether the current configuration is valid
 func (c *Config) Validate() error {
 	// #nosec G204
-	if _, err := exec.Command(c.Distribution.Go, "env").CombinedOutput(); err != nil {
-		path, err := exec.LookPath("go")
-		if err != nil {
-			return ErrGoNotFound
+	if !(c.SkipCompilation && c.SkipGetModules) {
```
Comment (dmathieu) [inline cmd/builder/internal/builder/config.go:93]: This change seems unrelated to the original intent of your PR. How about splitting it into another one?

Also, a global negative condition not the most readable way to write this IMHO.
I would write this `if !c.SkipCompilation || !c.SkipGetModules`.

**Review by bogdandrutu — COMMENTED** [review]
To be honest, I would be more aggressive into looking to remove the `go mod tidy` step and compilation. We should just generate the code from this tool. The other operations are very dependent on the environment.

**Comment by jpkrohling** [comment]
> To be honest, I would be more aggressive into looking to remove the go mod tidy step and compilation

No, this goes against the spirit of the tool. Give it a manifest, and it will give you a collector distribution binary. I'm OK with having flags to fine-tune this and have the builder generate only sources if needed, but the tool should be able to generate binaries.

**Comment by bogdandrutu** [comment]
> No, this goes against the spirit of the tool. Give it a manifest, and it will give you a collector distribution binary. I'm OK with having flags to fine-tune this and have the builder generate only sources if needed, but the tool should be able to generate binaries.

Then we should split the tool into 2, generator and builder :) I feel that the build part is a well covered tooling problem, only the generation is specific to the collector.

**Comment by jpkrohling** [comment]
Why have separate tools when one flag can be used to skip the compilation step?

**Comment by bogdandrutu** [comment]
> Why have separate tools when one flag can be used to skip the compilation step?

The reason for me is that, I believe that the tooling for go mods and go builds are there, and don't trust too much "magic" behind. The skip is fine, but I don't like how complicated logic we have to determine the go binary, etc.

**Comment by bogdandrutu** [comment]
@jpkrohling anyway, having a flag for the moment is fine, I am just saying that we should probably not do this in the tool, but can have that discussion separately.

File: CHANGELOG.md (hunk 1 above)
Comment (foadnh) [reply CHANGELOG.md:14]: Thanks, fixed it.

File: cmd/builder/internal/builder/config.go (hunk 2 above)
Comment (foadnh) [reply cmd/builder/internal/builder/config.go]: Good point. I will move it to a separate PR and create a new one once this PR is merged.

**Review by jpkrohling — APPROVED** [review]
